cache/cache.go

```diff
@@ -29,129 +31,181 @@ type LoadingCache interface {
+			return nil, errors.Wrap(err, "failed to set cache option")
 		}
 	}
 
+	if res.ttl == 0 {
+		res.ttl = noEvictionTTL // very long ttl to prevent eviction
+	}
+
 	if res.maxKeys > 0 || res.purgeEvery > 0 {
 		if res.purgeEvery == 0 {
 			res.purgeEvery = time.Minute * 5 // non-zero purge enforced because maxKeys defined
 		}
-		go func() {
+		go func(done <-chan struct{}) {
 			ticker := time.NewTicker(res.purgeEvery)
-			for range ticker.C {
-				res.Lock()
-				res.purge(res.maxKeys)
-				res.Unlock()
+			for {
+				select {
+				case <-done:
+					return
+				case <-ticker.C:
+					res.Lock()
+					res.purge(res.maxKeys)
+					res.Unlock()
+				}
 			}
-		}()
+		}(res.done)
 	}
-	log.Printf("[DEBUG] cache created. purge=%v, max=%d", res.purgeEvery, res.maxKeys)
-	return &res
+	return &res, nil
 }
 
-// Get by key if found, or get value and return
-func (c *loadingCacheImpl) Get(key string, ttl time.Duration, f func() (interface{}, error)) (interface{}, error) {
-
+// Set key
+func (c *loadingCacheImpl) Set(key string, value interface{}) {
 	c.Lock()
-	var ci *cacheItem
-	if ci = c.data[key]; ci == nil {
-
-		itemTTL := ttl
-		if ttl == 0 {
-			itemTTL = noEvictionTTL // very long ttl to prevent eviction
-		}
+	defer c.Unlock()
 
-		ci = &cacheItem{fun: f, ttl: itemTTL}
-		c.data[key] = ci
-		c.stat.Added++
+	now := time.Now()
+	_, ok := c.data[key]
+	if !ok {
+		c.data[key] = &cacheItem{}
+	}
+	c.data[key].data = value
+	c.data[key].expiresAt = now.Add(c.ttl)
+	if c.isLRU {
+		c.data[key].lastReadAt = now
 	}
-	c.Unlock()
 
-	// we don't want to block the possible call as this may take time
-	// and also cause dead-lock if one cached value calls (the same cache) for another
-	data, called, err := ci.value(c.allowError)
+	if c.maxKeys > 0 && int64(len(c.data)) >= c.maxKeys*2 {
+		c.purge(c.maxKeys)
+	}
+}
 
+// Get returns the key value
+func (c *loadingCacheImpl) Get(key string) (interface{}, bool) {
 	c.Lock()
 	defer c.Unlock()
-	c.stat.Size = len(c.data)
-	if called {
-		c.stat.Misses++
-	} else {
-		c.stat.Hits++
+	value, ok := c.getValue(key)
+	if !ok {
+		return nil, false
 	}
-	if c.maxKeys > 0 && c.stat.Size >= c.maxKeys*2 {
-		c.purge(c.maxKeys)
+	if c.isLRU {
+		c.data[key].lastReadAt = time.Now()
 	}
-	return data, err
+	return value, ok
 }
 
-// GetValue by key from cache
-func (c *loadingCacheImpl) GetValue(key string) (interface{}, bool) {
+// Peek returns the key value (or undefined if not found) without updating the "recently used"-ness of the key.
+func (c *loadingCacheImpl) Peek(key string) (interface{}, bool) {
 	c.Lock()
 	defer c.Unlock()
-	res, ok := c.data[key]
+	value, ok := c.getValue(key)
 	if !ok {
 		return nil, false
 	}
-	return res.data, ok
+	return value, ok
 }
 
 // Invalidate key (item) from the cache
 func (c *loadingCacheImpl) Invalidate(key string) {
 	c.Lock()
-	delete(c.data, key)
+	if value, ok := c.data[key]; ok {
+		if c.onEvicted != nil {
+			c.onEvicted(key, value)
+		}
+		delete(c.data, key)
```

Comment:
do we need to call onEvicted before or after the actual evictions?

Reply:
[gcache](https://github.com/bluele/gcache/blob/bc40bd6538339bd4be8cb0b48fc33fdf012ddb6e/lru.go#L221) calls it after, as well as [go-cache](https://github.com/patrickmn/go-cache/blob/master/cache.go#L910). I'll change our implementation to call it after as well.